text/0000-eager-macro-expansion.md

```diff
@@ -0,0 +1,702 @@
+1990](https://github.com/rust-lang/rfcs/pull/1990), indicating that macros in
+attributes could be used to solve problems at least important enough to go
+through the RFC process.
+
+## Interpolating macros in output
+
+Macros are currently not allowed in certain syntactic positions. Famously, they
+aren't allowed in identifier position, which makes `concat_idents!` [almost
+useless](https://github.com/rust-lang/rust/issues/29599). If macro authors have
+access to eager expansion, they could eagerly expand `concat_idents!` and
+interpolate the resulting token into their output.
+
+# Guide-level explanation
+
+We expose the following API in the `proc_macro` crate, to allow proc macro
+authors to iteratively expand all macros within a token stream:
+
+```rust
+use proc_macro::{LexError, Diagnostic, TokenStream};
+
+pub struct ExpansionBuilder { /* No public fields. */ }
+pub enum ExpansionError {
+    LexError(LexError),
+    ParseError(Diagnostic),
+    MissingDefinition(Diagnostic),
+    /* What other errors do we want to report?
+       What failures can the user reasonably handle? */
+}
+
+impl ExpansionBuilder {
+    /// Creates a new macro expansion request to expand macro invocations that
+    /// occur in `tokens`.
+    ///
+    /// Expansion results will be interpolated within the input stream before
+    /// being returned.
+    ///
+    /// By default, macro invocations in `tokens` will have their expansion
+    /// results recursively expanded, until there are no invocations left to
+    /// expand. To change this, use the `set_max_depth` method on the returned
+    /// instance of `ExpansionBuilder`.
+    ///
+    /// `tokens` should parse as valid Rust -- for instance, as an item or
+    /// expression.
+    pub fn from_tokens(tokens: TokenStream) -> Self;
+
+    /// Sends the expansion request to the compiler, then awaits the results of
+    /// expansion.
+    ///
+    /// The returned future will be ready unless expansion requires expanding a
+    /// procedural macro or a macro that hasn't been defined yet but might be
+    /// after further expansion. In those cases, the returned future will be
+    /// woken once all the required expansions have completed.
+    pub async fn expand(self) -> Result<TokenStream, ExpansionError>;
+
+    /// Sets the maximum depth of expansion. For example, if the depth is 2,
+    /// then the result of expanding the following tokens:
+    /// ```rust
+    /// {
+    ///     vec![vec![vec![17; 1]; 2]; 3];
+    ///     concat!("a", "b");
+    /// }
+    /// ```
+    /// Will be the following tokens:
+    /// ```rust
+    /// {
+    ///     std::vec::from_elem(std::vec::from_elem(vec![17; 1]. 2), 3);
+    ///     "ab";
+    /// }
+    /// ```
+    ///
+    /// Notice that, since the innermost invocation of `vec!` is "inside" two
+    /// other invocatoins of `vec!`, it is left unexpanded.
+    ///
+    /// If `depth` is 0, all macro invocations in the input will be recursively
```

Comment:
Bikeshedding: what's the idiomatic way to have a special case for an option like this? Tempted to make the argument an `Option<usize>`.

Reply:
Should there be an unbounded macro expansion at all?

Reply:
If I were a user of this API in any of the situations outlined under 'motivations', I would want to do an unbounded expansion, and if that wasn't an option I would use `set_max_depth(usize::MAX)`.

Reply:
We probably should have a warning after a certain number of seconds elapsed during expansion (say after one minute?) anyway, so fair enough.